README
NAME
    `Tickit' - Terminal Interface Construction KIT

SYNOPSIS
     use Tickit;
     use Tickit::Widget::Box;
     use Tickit::Widget::Static;

     my $box = Tickit::Widget::Box->new(
        h_border => 4,
        v_border => 2,
        bg       => "green",
        child    => Tickit::Widget::Static->new(
           text     => "Hello, world!",
           bg       => "black",
           align    => "centre",
           valign   => "middle",
        ),
     );

     Tickit->new( root => $box )->run;

DESCRIPTION
    `Tickit' is a high-level toolkit for creating full-screen terminal-based
    interactive programs. It allows programs to be written in an abstracted
    way, working with a tree of widget objects, to represent the layout of
    the interface and implement its behaviours.

    Its supported terminal features includes a rich set of rendering
    attributes (bold, underline, italic, 256-colours, etc), support for
    mouse including wheel and position events above the 224th column and
    arbitrary modified key input via libtermkey (all of these will require a
    supporting terminal as well). It also supports having multiple instances
    and non-blocking or asynchronous control.

    At the current version, this is a Perl distribution which contains and
    XS and C implementation of the lower levels (Tickit::Term and
    Tickit::Pen), and implements the higher levels (Tickit::Window and
    Tickit::Widget) in pure perl. The XS parts are supported by libtickit,
    either from the installed library, or using a bundled copy compiled at
    build time. It is intended that eventually the Window layer will be
    rewritten in XS and C instead.

CONSTRUCTOR
  $tickit = Tickit->new( %args )
    Constructs a new `Tickit' framework container object.

    Takes the following named arguments at construction time:

    term_in => IO
            IO handle for terminal input. Will default to `STDIN'.

    term_out => IO
            IO handle for terminal output. Will default to `STDOUT'.

    UTF8 => BOOL
            If defined, overrides locale detection to enable or disable
            UTF-8 mode. If not defined then this will be detected from the
            locale by using Perl's `${^UTF8LOCALE}' variable.

    root => Tickit::Widget
            If defined, sets the root widget using `set_root_widget' to the
            one specified.

METHODS
  $tickit->later( $code )
    Runs the given CODE reference at some time soon in the future. It will
    not be invoked yet, but will be invoked at some point before the next
    round of input events are processed.

  $tickit->timer( $mode, $amount, $code )
    Runs the given CODE reference at some fixed point in time in the future.
    `$mode' must be either the string `at', or `after'; and specifies that
    `$amount' gives either the absolute epoch time, or the delay relative to
    now, respectively. Fractions are supported to a resolution of
    microseconds.

     $tickit->timer( at => $epoch, $code )

     $tickit->timer( after => $delay, $code )

  $term = $tickit->term
    Returns the underlying Tickit::Term object.

  $cols = $tickit->cols
  $lines = $tickit->lines
    Query the current size of the terminal. Will be cached and updated on
    receipt of `SIGWINCH' signals.

  $tickit->bind_key( $key, $code )
    Installs a callback to invoke if the given key is pressed, overwriting
    any previous callback for the same key. The code block is invoked as

     $code->( $tickit, $key )

    If `$code' is missing or `undef', any existing callback is removed.

    As a convenience for the common application use case, the `Ctrl-C' key
    is bound to the `stop' method.

    To remove this binding, simply bind another callback, or remove the
    binding entirely by setting `undef'.

  $tickit->rootwin
    Returns the root Tickit::Window.

  $tickit->set_root_widget( $widget )
    Sets the root widget for the application's display. This must be a
    subclass of Tickit::Widget.

  $tickit->setup_term
    Set up the screen and generally prepare to start running

  $tickit->teardown_term
    Shut down the screen after running

  $tickit->tick
    Run a single round of IO events. Does not call `setup_term' or
    `teardown_term'.

  $tickit->run
    Calls the `setup_term' method, then processes IO events until stopped,
    by the `stop' method, `SIGINT', `SIGTERM' or the `Ctrl-C' key. Then runs
    the `teardown_term' method, and returns.

  $tickit->stop
    Causes a currently-running `run' method to stop processing events and
    return.

AUTHOR
    Paul Evans <leonerd@leonerd.org.uk>



Changes
Revision history for Changes

0.49    2014/12/16 21:11:15
        [CHANGES]
         * Perform 0.49 deprecations:
            + Announce deprecation of Tickit::Pen observers
            + Warn on Widgets that don't have WIDGET_PEN_FROM_STYLE set
            + Die on attempts to disable CHILD_WINDOWS_LATER
         * Allow chaining return of ContainerWidget->{add,remove}

0.48    2014/09/18 20:55:38
        [CHANGES]
         * Perform 0.48 removal deprecations
         * Added Tickit::Test::clear_term
         * Print deprecation warnings from Tickit::Window direct draw methods

        [BUGFIXES]
         * Test for presence of T:W:HBox at BEGIN time during unit test, to
           avoid error
         * Fix for Tickit::Utils::distribute() with fixed=0 when in
           overallocation mode
         * Ensure that child window on_expose handler is run before containing
           parent, to ensure RB holes are set correctly for underprint
         * Hidden child windows should not count during $win->scroll*
         * Ensure child window key handler propagation doesn't continue after
           a key event has been handled

0.47    2014/09/05 23:22:29
        [CHANGES]
         * Moved Tickit::Widget::HBox/VBox to the Tickit-Widgets dist
         * Remove Tickit::WidgetRole::Penable
         * Support the new 'blink' Pen attribute

        [BUGFIXES]
         * Bundled libtickit bugfixes:
            + Improved handling of some missing terminfo attributes
            + Respect renderbuffer offset and limits in get_* functions

0.46    2014/05/02 20:16:37
        [CHANGES]
         * Move RenderBuffer and MockTerm entirely into libtickit C library
         * Added $rb->char
         * Add Tickit::Debug support to RenderBuffer

        [BUGFIXES]
         * Ensure that $rb->skip/skip_at obey masking regions

0.45    2014/04/15 02:47:45
        [CHANGES]
         * Move RenderBuffer's linemask-to-character conversion code into XS/C
           to prepare for RB-in-C
         * Implement most of Tickit::Test::MockTerm logic in C
         * Don't emit INSERTCH or DELETECH events from MockTerm
         * Announce $rb->flush_to_window as deprecated
         * Announce Window direct drawing methods as deprecated
         * Emit deprecation warnings from $term->mode_* and $win->*penattrs
           methods
         * Remove the Tickit::WidgetRole::Penable code; replace with a simple
           die message

0.44    2014/04/07 02:03:37
        [CHANGES]
         * Deprecation warnings for pre-0.42-style Window expose/key/mouse
           event handlers
         * Provide $ev->mod_is_{shift,ctrl,alt} shortcut tests
         * Added $rb->linebox_at()
         * Allow ContainerWidgets to provide a method to return only those
           children it wants to focus search on
         * Added Tickit::Debug logging
         * Added focus_child_notify Window ability and :focus-child
           ContainerWidget style tag
         * Use a single RenderBuffer for the entire window/widget tree
           rendering
         * Added $rb->get_cell accessor for inspecting cell state
        
        [BUGFIXES]
         * Ensure that focus searching ability only considers children with
           visible windows
         * Ensure that $rb->mask won't crash with over-sized or negative-
           offset mask rects
         * Clip $win->scroll rect to the actual window bounds
         * Correct explanation of tagged vs. untagges style in docs
         * Fix $win->scrollrect when moving chunks a distance larger than
           their own size
         * Fix $rb->DESTROY with content to not print warnings of unreferenced
           scalars

0.43    2014/04/01 16:29:58
        [CHANGES]
         * Default TICKIT_CHILD_WINDOWS_LATER on; with env. var to disable it
         * Allow optional $pen argument to $term->{print,clear,erasech}
         * Support flushing of RenderBuffer directly to a Tickit::Term
         * Store pending 'later' geometry changes separately from expose logic
           to ensure correct output
         * Rewrite of Window rendering logic:
            + accumulate damage and run expose logic only on root window
         * Provide a more robust associative version of is_termlog() to avoid
           test logic that depends on exact rendering order
        
        [BUGFIXES]
         * Ensure that $win->subwindows works on windows with no subs
         * $win->restore should remember to flush the actual terminal buffer
         * Ensure that RB masks are correctly interpreted during translations

0.42    2014/01/12 13:40:39
        [CHANGES]
         * Added Window on_expose callback type with RenderBuffer
         * Added legacy-detecting $win->set_on_* methods to set new-style
           Window callbacks
         * Perform Widget rendering onto Window using new with_rb expose event
         * Deprecate legacy versions of Window on_expose, on_key, on_mouse
         * Removed CLEAR_BEFORE_RENDER

0.41    2013/12/19 19:17:07
        [CHANGES]
         * Allow cursors to be entirely hidden on a Window
         * Return $text width in columns from $rb->text and $rb->text_at
         * Document that Tickit::Style->load_style* are methods, not functions
           (RT89235)

        [BUGFIXES]
         * Fix for T:W:Box - set 'dir' of T:WR:Alignable, not 'style'
         * Fix double-nested container focus management

0.40    2013/09/29 19:08:09
        [CHANGES]
         * New size model; ->requested_{size,lines,cols} which cache
         * Lazy $parent->resize only when required
         * Updated all examples/docs for new ->render_to_rb
         * Print a legacy deprecation warning when constructing a
           ->render-based widget
         * Allow Styled widgets to copy their base class
         * Allow HBox/VBox to take a 'children' ARRAY to constructor

        [BUGFIXES]
         * Specify minimum required version of Alien::libtermkey to get
           libtermkey >= 0.16

lib/Tickit.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2009-2014 -- leonerd@leonerd.org.uk

package Tickit;

use strict;
use warnings;

BEGIN {
   our $VERSION = '0.49';
}

use Carp;

use IO::Handle;

use Scalar::Util qw( weaken );
use Time::HiRes qw( time );

BEGIN {
   require XSLoader;
   XSLoader::load( __PACKAGE__, our $VERSION );
}

use Tickit::Term qw( TERM_MOUSEMODE_DRAG );
use Tickit::Window;

use Tickit::Debug;

use Struct::Dumb;
struct TimeQueue => [qw( time code )];

=head1 NAME

C<Tickit> - Terminal Interface Construction KIT

=head1 SYNOPSIS

 use Tickit;
 use Tickit::Widget::Box;
 use Tickit::Widget::Static;

 my $box = Tickit::Widget::Box->new(
    h_border => 4,
    v_border => 2,
    bg       => "green",
    child    => Tickit::Widget::Static->new(
       text     => "Hello, world!",
       bg       => "black",
       align    => "centre",
       valign   => "middle",
    ),
 );

 Tickit->new( root => $box )->run;

=head1 DESCRIPTION

C<Tickit> is a high-level toolkit for creating full-screen terminal-based
interactive programs. It allows programs to be written in an abstracted way,
working with a tree of widget objects, to represent the layout of the
interface and implement its behaviours.

Its supported terminal features includes a rich set of rendering attributes
(bold, underline, italic, 256-colours, etc), support for mouse including wheel
and position events above the 224th column and arbitrary modified key input
via F<libtermkey> (all of these will require a supporting terminal as well).
It also supports having multiple instances and non-blocking or asynchronous
control.

At the current version, this is a Perl distribution which contains and XS and
C implementation of the lower levels (L<Tickit::Term> and L<Tickit::Pen>), and
implements the higher levels (L<Tickit::Window> and L<Tickit::Widget>) in pure
perl. The XS parts are supported by F<libtickit>, either from the installed
library, or using a bundled copy compiled at build time. It is intended that
eventually the Window layer will be rewritten in XS and C instead.

=cut

=head1 CONSTRUCTOR

=cut

=head2 $tickit = Tickit->new( %args )

Constructs a new C<Tickit> framework container object.

Takes the following named arguments at construction time:

=over 8

=item term_in => IO

IO handle for terminal input. Will default to C<STDIN>.

=item term_out => IO

IO handle for terminal output. Will default to C<STDOUT>.

=item UTF8 => BOOL

If defined, overrides locale detection to enable or disable UTF-8 mode. If not
defined then this will be detected from the locale by using Perl's
C<${^UTF8LOCALE}> variable.

=item root => Tickit::Widget

If defined, sets the root widget using C<set_root_widget> to the one
specified.

=back

=cut

sub new
{
   my $class = shift;
   my %args = @_;

   # Test code also accepts 'term' argument but we won't document that for now

   my $in  = delete $args{term_in}  || \*STDIN;
   my $out = delete $args{term_out} || \*STDOUT;

   my $term = delete $args{term};

   my $self = bless {
      todo_queue => [],
      timer_queue => [],
      redraw_queue => [],
   }, $class;

   unless( $term ) {
      my $writer = $self->_make_writer( $out );

      $term = Tickit::Term->find_for_term(
         writer        => $writer,
         input_handle  => $in,
         output_handle => $out,
         UTF8          => $args{UTF8},
      );
   }

   $self->{term} = $term;
   $self->{term_in}  = $in;
   $self->{term_out} = $out;

   my $rootwin = $self->{rootwin} = Tickit::Window->new( $self, $term->lines, $term->cols );

   $self->bind_key( 'C-c' => $self->can( "stop" ) );

   $term->set_output_buffer( 2**16 ); # 64KiB

   weaken( my $weakself = $self );

   $term->bind_event( resize => sub {
      $weakself or return;
      my ( $term, $ev, $args ) = @_;
      Tickit::Debug->log( Ir => "Resize to $args->{cols}x$args->{lines}" ) if DEBUG;
      $weakself->rootwin->resize( $args->{lines}, $args->{cols} );
   } );

   $term->bind_event( key => sub {
      $weakself or return;
      my ( $term, $ev, $args ) = @_;
      Tickit::Debug->log( Ik => "Key event %s %s (mod=%d)", @{$args}{qw( type str mod )} ) if DEBUG;
      $weakself->on_key( $args );
   } );

   $term->bind_event( mouse => sub {
      $weakself or return;
      my ( $term, $ev, $args ) = @_;
      Tickit::Debug->log( Im => "Mouse event %s => %s @%d,%d (mod=%d)", @{$args}{qw( type button col line mod )} ) if DEBUG;
      $weakself->on_mouse( $args );
   } );

   $self->set_root_widget( $args{root} ) if defined $args{root};

   return $self;
}

=head1 METHODS

=cut

sub _make_writer
{
   my $self = shift;
   my ( $out ) = @_;

   $out->autoflush( 1 );

   return $out;
}

=head2 $tickit->later( $code )

Runs the given CODE reference at some time soon in the future. It will not be
invoked yet, but will be invoked at some point before the next round of input
events are processed.

=cut

sub _flush_later
{
   my $self = shift;

   my $queue = $self->{todo_queue};
   ( shift @$queue )->() while @$queue;
}

sub later
{
   my $self = shift;
   my ( $code ) = @_;

   push @{ $self->{todo_queue} }, $code;
}

=head2 $tickit->timer( $mode, $amount, $code )

Runs the given CODE reference at some fixed point in time in the future.
C<$mode> must be either the string C<at>, or C<after>; and specifies that
C<$amount> gives either the absolute epoch time, or the delay relative to now,
respectively. Fractions are supported to a resolution of microseconds.

 $tickit->timer( at => $epoch, $code )

 $tickit->timer( after => $delay, $code )

=cut

sub timer
{
   my $self = shift;
   my ( $mode, $amount, $code ) = @_;

   my $at;
   if( $mode eq "at" ) {
      $at = $amount;
   }
   elsif( $mode eq "after" ) {
      $at = time + $amount;
   }
   else {
      croak "Mode should be 'at' or 'after'";
   }

   # TODO: bin-search insert position then splice
   @{ $self->{timer_queue} } = sort { $a->time <=> $b->time } @{ $self->{timer_queue} }, TimeQueue( $at => $code );
}

=head2 $term = $tickit->term

Returns the underlying L<Tickit::Term> object.

=cut

sub term
{
   my $self = shift;
   return $self->{term};
}

=head2 $cols = $tickit->cols

=head2 $lines = $tickit->lines

Query the current size of the terminal. Will be cached and updated on receipt
of C<SIGWINCH> signals.

=cut

sub lines { shift->term->lines }
sub cols  { shift->term->cols  }

sub _SIGWINCH
{
   my $self = shift;

   $self->term->refresh_size;
}

sub on_key
{
   my $self = shift;
   my ( $args ) = @_;

   my $str = $args->{str};

   $self->rootwin->_handle_key( $args ) and return;

   if( exists $self->{key_binds}{$str} ) {
      $self->{key_binds}{$str}->( $self, $str ) and return;
   }
}

=head2 $tickit->bind_key( $key, $code )

Installs a callback to invoke if the given key is pressed, overwriting any
previous callback for the same key. The code block is invoked as

 $code->( $tickit, $key )

If C<$code> is missing or C<undef>, any existing callback is removed.

As a convenience for the common application use case, the C<Ctrl-C> key is
bound to the C<stop> method.

To remove this binding, simply bind another callback, or remove the binding
entirely by setting C<undef>.

=cut

sub bind_key
{
   my $self = shift;
   my ( $key, $code ) = @_;

   if( $code ) {
      $self->{key_binds}{$key} = $code;
   }
   else {
      delete $self->{key_binds}{$key};
   }
}

sub on_mouse
{
   my $self = shift;
   my ( $args ) = @_;

   my $type = $args->{type};
   if( $type eq "press" ) {
      # Save the mouse press location in case of drag
      $self->{mouse_last_press} = [ $args->{button}, $args->{line}, $args->{col} ];
   }
   elsif( $type eq "drag" and !$self->{mouse_dragging} ) {
      my $start_args = { %$args,
         type => "drag_start",
         line => $self->{mouse_last_press}[1],
         col  => $self->{mouse_last_press}[2],
      };

      my $win = $self->rootwin->_handle_mouse( $start_args );
      $self->{drag_source_window} = $win;

      $self->{mouse_dragging} = 1;
   }
   elsif( $type eq "release" and $self->{mouse_dragging} ) {
      my $drop_args = { %$args,
         type => "drag_drop",
      };

      $self->rootwin->_handle_mouse( $drop_args );

      if( my $win = $self->{drag_source_window} ) {
         $win->_handle_mouse( { %$args,
            type => "drag_stop",
            line => $args->{line} - $win->abs_top,
            col  => $args->{col}  - $win->abs_left,
         } );
      }

      $self->{mouse_dragging} = 0;
   }

   my $win = $self->rootwin->_handle_mouse( $args );

   my $srcwin;
   if( $type eq "drag" and
       $srcwin = $self->{drag_source_window} and
       !$win || ( $win && $win != $srcwin ) ) {
      $srcwin->_handle_mouse( { %$args,
         type => "drag_outside",
         line => $args->{line} - $srcwin->abs_top,
         col  => $args->{col}  - $srcwin->abs_left,
      } );
   }
}

=head2 $tickit->rootwin

Returns the root L<Tickit::Window>.

=cut

sub rootwin
{
   my $self = shift;
   return $self->{rootwin};
}

=head2 $tickit->set_root_widget( $widget )

Sets the root widget for the application's display. This must be a subclass of
L<Tickit::Widget>.

=cut

sub set_root_widget
{
   my $self = shift;
   ( $self->{root_widget} ) = @_;
}

=head2 $tickit->setup_term

Set up the screen and generally prepare to start running

=cut

sub setup_term
{
   my $self = shift;

   my $term = $self->term;

   $term->await_started( 0.100 ); # 100 msec

   $term->setctl_int( altscreen => 1 );
   $term->setctl_int( cursorvis => 0 );
   $term->setctl_int( mouse     => TERM_MOUSEMODE_DRAG );
   $term->clear;

   if( my $widget = $self->{root_widget} ) {
      $widget->set_window( $self->rootwin );
   }

   $term->flush;
}

=head2 $tickit->teardown_term

Shut down the screen after running

=cut

sub teardown_term
{
   my $self = shift;

   if( my $widget = $self->{root_widget} ) {
      $widget->set_window( undef );
   }

   my $term = $self->term;
   $term->setctl_int( altscreen => 0 );
   $term->setctl_int( cursorvis => 1 );
   $term->setctl_int( mouse     => 0 );

   $term->flush;
}

=head2 $tickit->tick

Run a single round of IO events. Does not call C<setup_term> or
C<teardown_term>.

=cut

sub _tick
{
   my $self = shift;

   my $timer_queue = $self->{timer_queue};

   my $timeout;
   if( @$timer_queue ) {
      $timeout = $self->{timer_queue}[0]->time - time;
   }

   $self->{term}->input_wait( $timeout );

   my $now = time;
   while( @$timer_queue and $timer_queue->[0]->time <= $now ) {
      shift( @$timer_queue )->code->();
   }

   $self->_flush_later if @{ $self->{todo_queue} };
}

sub tick
{
   my $self = shift;

   my $old_DIE = $SIG{__DIE__};
   local $SIG{__DIE__} = sub {
      local $SIG{__DIE__} = $old_DIE;

      die @_ if $^S;

      $self->teardown_term;
      die @_;
   };

   $self->_tick;
}

=head2 $tickit->run

Calls the C<setup_term> method, then processes IO events until stopped, by the
C<stop> method, C<SIGINT>, C<SIGTERM> or the C<Ctrl-C> key. Then runs the
C<teardown_term> method, and returns.

=cut

sub run
{
   my $self = shift;

   $self->setup_term;

   $SIG{INT} = $SIG{TERM} = sub {
      my $signal = shift;

      # Disable the handler so a second signal will be fatal
      undef $SIG{$signal};
      $SIG{$signal} = sub {
         die "Aborting Tickit on double SIGINT\n";
      };

      $self->stop;
   };

   $SIG{WINCH} = sub {
      $self->later( sub { $self->_SIGWINCH } )
   };

   my $old_DIE = $SIG{__DIE__};
   local $SIG{__DIE__} = sub {
      local $SIG{__DIE__} = $old_DIE;

      die @_ if $^S;

      $self->teardown_term;
      die @_;
   };

   $self->_flush_later if @{ $self->{todo_queue} };

   local $self->{keep_running} = 1;
   $self->_tick while( $self->{keep_running} );

   $self->teardown_term;
}

=head2 $tickit->stop

Causes a currently-running C<run> method to stop processing events and return.

=cut

sub stop
{
   my $self = shift;
   $self->{keep_running} = 0;
}

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;
